academy/ibc-dev/ibcapp_checkers.md

```diff
@@ -0,0 +1,525 @@
+# Extend the Checkers Game With a Leaderboard
+
+Go to your checkers folder and make sure that you checked out of the [v1-cosmjs-elements](https://github.com/cosmos/b9-checkers-academy-draft/tree/v1-cosmjs-elements) tag.
+
+In the checkers chain folder, you can scaffold a leaderboard module with Ignite:
+
+```bash
+$ ignite scaffold module leaderboard --ibc
+```
+
+In order to create and maintain a leaderboard, you need to store the player information.
+
+Scaffold a structure with:
+
+```bash
+$ ignite scaffold map playerInfo wonCount:uint lostCount:uint dateUpdated:string --module leaderboard --no-message
+```
+
+Now you can use this structure to create the board itself:
+
+```bash
+$ ignite scaffold single board PlayerInfo:PlayerInfo --module leaderboard --no-message
+```
+
+You want the structures as [nullable types](https://en.wikipedia.org/wiki/Nullable_type), so a few adjustments are needed - especially because you do not have a null value for an address.
+
+You need to do the adjustments in the Protobuf files `proto/leaderboard/board.proto` and `proto/leaderboard/genesis.proto`. Make sure to import `gogoproto/gogo.proto` and use `[(gogoproto.nullable) = false];` for the `PlayerInfo` and the `Board`, like:
+
+```protobuf
+syntax = "proto3";
+package b9lab.checkers.leaderboard;
+
+option go_package = "github.com/b9lab/checkers/x/leaderboard/types";
+import "leaderboard/player_info.proto"; 
+import "gogoproto/gogo.proto";
+
+message Board {
+  repeated PlayerInfo playerInfo = 1 [(gogoproto.nullable) = false]; 
+  
+}
+```
+
+for `proto/leaderboard/board.proto`.
+
+You want to store a win, a loss, or a draw if a game ends. Therefore, you should create some helper functions first. Create a `x/checkers/keeper/player_info_handler.go` file with the following code:
+
+```golang
+package keeper
+
+import (
+    "fmt"
+
+    rules "github.com/b9lab/checkers/x/checkers/rules"
+    "github.com/b9lab/checkers/x/checkers/types"
+
+    sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func getWinnerAndLoserAddresses(storedGame *types.StoredGame) (winnerAddress sdk.AccAddress, loserAddress sdk.AccAddress) {
+    if storedGame.Winner == rules.PieceStrings[rules.NO_PLAYER] {
+        panic(types.ErrThereIsNoWinner.Error())
+    }
+    redAddress, err := storedGame.GetRedAddress()
+    if err != nil {
+        panic(err.Error())
+    }
+    blackAddress, err := storedGame.GetBlackAddress()
+    if err != nil {
+        panic(err.Error())
+    }
+    if storedGame.Winner == rules.PieceStrings[rules.RED_PLAYER] {
+        winnerAddress = redAddress
+        loserAddress = blackAddress
+    } else if storedGame.Winner == rules.PieceStrings[rules.BLACK_PLAYER] {
+        winnerAddress = blackAddress
+        loserAddress = redAddress
+    } else {
+        panic(fmt.Sprintf(types.ErrWinnerNotParseable.Error(), storedGame.Winner))
```

Comment:
Given comment above, I started from another commit so this might explain, but I had to register this error `ErrWinnerNotParseable`

Reply:
Good catch, was missing in the correct commit too.